### PR TITLE
Improve detection of incorrect lines during gravity

### PR DESCRIPTION
**Pull request by yubiuser**
**What does this PR aim to accomplish?:**

When `gravity` downloads a file we use a two-step "cleaning" mechanism (`gravity_ParseFileIntoDomains` and `parseList`).
https://github.com/pi-hole/pi-hole/blob/e773e3302ca66a6d918a40c8a8c6282f223d4906/gravity.sh#L686-L688

In `gravity_ParseFileIntoDomains` we already remove lines which don't match domain names 
https://github.com/pi-hole/pi-hole/blob/e773e3302ca66a6d918a40c8a8c6282f223d4906/gravity.sh#L742

Later, in `parseList` we use a different regex to filter for invalid domains
https://github.com/pi-hole/pi-hole/blob/e773e3302ca66a6d918a40c8a8c6282f223d4906/gravity.sh#L533

However, we use the output of `gravity_ParseFileIntoDomains` to see if there are "incorrect_lines" in the file
https://github.com/pi-hole/pi-hole/blob/e773e3302ca66a6d918a40c8a8c6282f223d4906/gravity.sh#L535

Due to the cleaning already in `gravity_ParseFileIntoDomains` the number of `incorrect_lines` can be underestimated.
This is reported here: https://github.com/pi-hole/pi-hole/issues/5008

**How does this PR accomplish the above?:**

The regex from `gravity_ParseFileIntoDomains` is now part of `parseList`, meaning the "input" is less filtered (preventing underestimation). Within `parseList` both regex are used to filter incorrect lines separately (because they find different - although partly overlapping - incorrect lines), and sorted for unique entries afterwards.

---
**By submitting this pull request, I confirm the following:**

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against.
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
6. I have checked that another pull request for this purpose does not exist.
7. I have considered, and confirmed that this submission will be valuable to others.
8. I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
9. I give this submission freely, and claim no ownership to its content.

---
- [x] I have read the above and my PR is ready for review. *Check this box to confirm*

### CODE REVIEW

**Comment by PromoFaux** [comment]
Does the removed sed/regex work in the first place? Seems to just delete everything when I try on sed.js.org

image


The one we are left with appears to behave correctly except for happily accepting an empty line as a domain:

image

**Comment by yubiuser** [comment]
Don't omit the `-r` (regex-extended). For `sed.js.org` use the "option" menu.

**Comment by PromoFaux** [comment]
Cool - I missed that! 

One thing to note, then is that the regex that this PR intends to remove will mean that single words will get matched as domains (where they were removed before)

image


image

**Comment by yubiuser** [comment]
Ok, I see the problem. I'll think about a solution.

**Comment by yubiuser** [comment]
I improved the code: 

- `gravity_ParseFileIntoDomains` will now remove empty lines
- the regex from `gravity_ParseFileIntoDomains` is now part of `parseList`
- within `parseList` both regex are used to filter **incorrect** lines separately (because the find different - although partly overlapping - incorrect lines)
  - they are added to an array which is sorted for unique entries and limited to 5 examples

Because we enter `parseList` with a less filtered source the number of invalid lines reported by gravity will increase.

File: gravity.sh [hunk 1]
```diff
@@ -524,15 +524,20 @@ num_target_lines=0
 num_source_lines=0
 num_invalid=0
 parseList() {
-  local adlistID="${1}" src="${2}" target="${3}" incorrect_lines
+  local adlistID="${1}" src="${2}" target="${3}" incorrect_lines sample_incorrect_lines
   # This sed does the following things:
+  # 0. Remove all lines containing no domains
```
Comment (PromoFaux) [inline gravity.sh:529]: Can we make this a 1-based list rather than a 0-based one, please?

Comment (yubiuser) [reply gravity.sh:529]: Sure.

**Review by PromoFaux — APPROVED** [review]
Tested working in theory with with regex tester. Happy with your sample output too

**Pull request merged**